tests/groupRole.test.js

```diff
@@ -0,0 +1,1349 @@
+const ERR = require('async-stacktrace');
+
+const config = require('../lib/config');
+const path = require('path');
+var assert = require('chai').assert;
+var request = require('request');
+var cheerio = require('cheerio');
+
+var sqldb = require('@prairielearn/postgres');
+var sql = sqldb.loadSqlEquiv(__filename);
+
+var helperServer = require('./helperServer');
+
+let page, elemList;
+const locals = {};
+locals.helperClient = require('./helperClient');
+locals.siteUrl = 'http://localhost:' + config.serverPort;
+locals.baseUrl = locals.siteUrl + '/pl';
+locals.courseInstanceUrl = locals.baseUrl + '/course_instance/1';
+locals.assessmentsUrl = locals.courseInstanceUrl + '/assessments';
+locals.courseDir = path.join(__dirname, '..', 'testCourse');
+
+const storedConfig = {};
+
+describe('Test group based assessments with custom group roles from student side', function () {
+  this.timeout(20000);
+  before('set authenticated user', function (callback) {
+    storedConfig.authUid = config.authUid;
+    storedConfig.authName = config.authName;
+    storedConfig.authUin = config.authUin;
+    callback(null);
+  });
+  before('set up testing server', helperServer.before(locals.courseDir));
+  after('shut down testing server', helperServer.after);
+  after('unset authenticated user', function (callback) {
+    Object.assign(config, storedConfig);
+    callback(null);
+  });
+
+  describe('1. the database', function () {
+    it('should contain a group-based homework assessment with roles', function (callback) {
+      sqldb.query(sql.select_group_work_assessment_with_roles, [], function (err, result) {
+        if (ERR(err, callback)) return;
+        assert.lengthOf(result.rows, 1);
+        assert.notEqual(result.rows[0].id, undefined);
+        locals.assessment_id = result.rows[0].id;
+        locals.assessmentUrl = locals.courseInstanceUrl + '/assessment/' + locals.assessment_id;
+        callback(null);
+      });
+    });
+    it('should contain a group-based homework assessment without roles', function (callback) {
+      sqldb.query(sql.select_group_work_assessment_without_roles, [], function (err, result) {
+        if (ERR(err, callback)) return;
+        assert.lengthOf(result.rows, 1);
+        assert.notEqual(result.rows[0].id, undefined);
+        assert.equal(result.rows[0].has_roles, false);
+        locals.assessment_id_without_roles = result.rows[0].id;
+        locals.assessmentUrlWithoutRoles =
+          locals.courseInstanceUrl + '/assessment/' + locals.assessment_id_without_roles;
+        callback(null);
+      });
+    });
+    it('should contain the 4 roles for the assessment', function (callback) {
+      const params = {
+        assessment_id: locals.assessment_id,
+      };
+      sqldb.query(sql.select_assessment_group_roles, params, function (err, result) {
+        if (ERR(err, callback)) return;
+        assert.lengthOf(result.rows, 4);
```

Comment:
If you're going to use a giant shared object, you should use `mocha-steps` to ensure that if one step fails, the entire test aborts. See https://github.com/PrairieLearn/PrairieLearn/blob/master/tests/bonusPoints.test.js for an example. When you do that, you'll need to move all steps that depend on previous steps into the same `describe` block so that they can be aborted as a whole.